### PR TITLE
fix: detect newly added specs in dev-server compilation

### LINKED ISSUE #16664: Must restart Cypress to run newly added spec file when using Webpack 5
### Current behavior

If you add a new spec file after starting the component testing runner, eg `touch src/MySpec.spec.jsx`, it shows up in the runner, but it always says "no test found", even if you add some tests to the spec file.

You need to restart the runner to webpack to start watching the file correctly. 

It only happens for projects using Webpack 5.

### Desired behavior

Should work like it does for Webpack 4. I don't want to restart the runner.

### Test code to reproduce

- Get any webpack 5 repo, for example this one https://github.com/lmiller1990/cypress-react-webpack-5-example
- `yarn cypress open-ct`
- `touch src/Blah.spec.jsx`
- Click `Blah.spec.jsx` in the side bar
- Notice "no tests found"
- add a test, eg `it('something', () => {})`
- observe webpack does not notice the file changed or rebundle it

### Versions

- cypress 7.4.0
- latest cypress/webpack-dev-server
- webpack 5.x
- webpack dev server, probably 3.x

### PR DESCRIPTION
**Pull request by ZachJW34**
- Closes #16664

### Additional details
There are more fine-grained [API's](https://webpack.js.org/api/loaders/#thisaddcontextdependency) for controlling loaders. I tried a few, such as `addDependency` and `addContextDependency` without much luck. This loader is pretty cheap so I don't think disabling the caching would have much of a performance hit. Also, I still need to write a test covering this case.

### PR Tasks

- [x] Have tests been added/updated?
- [ ] Has the original issue or this PR been tagged with a release in ZenHub? 
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [ ] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

File: npm/webpack-dev-server/src/loader.ts [hunk 1]
```diff
@@ -49,7 +50,11 @@ function buildSpecs (projectRoot: string, files: Cypress.Cypress['spec'][] = [])
 }
 
 // Runs the tests inside the iframe
-export default function loader (this: CypressCTWebpackContext) {
+export default function loader (this: CypressCTWebpackContext & LoaderContext<any>) {
```
Comment (JessicaSachs) [inline npm/webpack-dev-server/src/loader.ts:53]: Don't you want `unknown`? CC @lmiller1990

Comment (lmiller1990) [reply npm/webpack-dev-server/src/loader.ts:53]: Actually we want `<void>` (or `{}` is probably fine), we are intentionally not passing any options to the loader context.

**Comment by lmiller1990** [comment]
This test might be close to what we need: https://github.com/cypress-io/cypress/blob/develop/npm/webpack-dev-server/test/e2e.spec.ts#L131

Basically when we make a new file, we want to ensure that webpack re-runs and emits the right event - that looks like what this test is doing.

File: npm/webpack-dev-server/src/loader.ts [hunk 2]
```diff
@@ -49,7 +50,11 @@ function buildSpecs (projectRoot: string, files: Cypress.Cypress['spec'][] = [])
...
+  // In Webpack 5, a spec added after the dev-server is created won't
+  // be included in the compilation. Disabling the caching of this loader ensures
+  // we regenerate our specs and include any new ones in the compilation.
+  this.cacheable(false)
```
Comment (lmiller1990) [inline npm/webpack-dev-server/src/loader.ts:57]: Do we want this to be conditional (if we are using Webpack 5)? Does `this.cacheable` even exist in Webpack 4?

Comment (JessicaSachs) [reply npm/webpack-dev-server/src/loader.ts:57]: It won't throw even if it doesn't. I'm fine with this.

**Comment by ZachJW34** [comment]
@lmiller1990 I was able to get some performance metrics after looking into #17920. Here is what I found testing inside a Next repo using Webpack 5:
```
Cache Enabled
  - Cold Start (ms): [2584, 262, 275, 257, 279]
  - Code Change (ms): [55, 58, 62, 41, 44]
 
Cache Disabled ("cache: false" in webpack config):
  - Cold Start (ms): [2747, 2541, 2453, 2214, 2591]
  - Code Change (ms): [1140, 729, 640, 591, 528]
 
Loader Cache Disabled ("this.cacheable(false)" in loader):
  - Cold Start (ms): [2679, 261, 290, 300, 274]
  - Code Change (ms): [53, 58, 58, 62, 50]
```
"Cold start" was the time it took for Webpack to when running `npx cypress open-ct` and the "Code Change" was the time it took Webpack to recompile after changing a spec file


Adding the [cache: false](https://webpack.js.org/configuration/cache/#cache) option to the webpack.config.js fixes the spec detection issue but at a high performance cost. 
**Conclusion**: The current implementation (disabling the caching of the loader) has very little performance impact.

File: npm/webpack-dev-server/src/loader.ts (hunk 2 above)
Comment (ZachJW34) [reply npm/webpack-dev-server/src/loader.ts:57]: `this.cacheable` is valid for Webpack 4 and 5
https://v4.webpack.js.org/api/loaders/#thiscacheable

**Review by lmiller1990 — APPROVED** [review]

**Comment by lmiller1990** [comment]
Great! Still in draft - any outstanding changes @Zacharias3690 and can we merge?

**Comment by Zacharias3690** [comment]
@lmiller1990 Looks like you tagged the wrong Zach :) Think you were looking for @ZachJW34

**Comment by lmiller1990** [comment]
@Zacharias3690 updated tagging the correct person... sorry about that!

**Comment by ZachJW34** [comment]
@lmiller1990 fixed the tests by adding `--ignore-scripts` to the yarn installs for the npm specific tests. Should be good to merge now

**Review by lmiller1990 — APPROVED** [review]

**Review by JessicaSachs — APPROVED** [review]

**Pull request merged**